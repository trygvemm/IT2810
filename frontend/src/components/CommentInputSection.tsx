import { gql, useMutation } from "@apollo/client";
import { useEffect, useState } from "react";
import CommentInput from "./types/CommentInput";
import CommentInputProps from "./types/CommentInputProps";

export const CREATE_COMMENT = gql`
	mutation createComment($text: String, $bookId: String) {
		createComment(input: { text: $text, bookId: $bookId }) {
			text
		}
	}
`;

export default function CommentInputSection(props: CommentInputProps) {
	const [commentText, setCommentText] = useState<string>("");
	const [show, setShow] = useState(false);

	// On componentDidMount set the timer
	useEffect(() => {
		const timeId = setTimeout(() => {
			// After 3 seconds set the show value to false
			setShow(false);
		}, 5000);

		return () => {
			clearTimeout(timeId);
		};
	}, [show]);

	const [createComment, { loading, error }] = useMutation<CommentInput>(CREATE_COMMENT, {
		refetchQueries: () => [
			{
				query: props.refetch,
				variables: {
					id: props.id,
				},
			},
		],
	});

	const postComment = () => {
		setShow(true);
		createComment({
			variables: { text: commentText, bookId: props.id },
		});
	};

	if (loading) return <p>Loading...</p>;
	if (error) {
		return <p>Error x(</p>;
	}

	return (
		<>
			<textarea
				id="commentTextArea"
				aria-label="Text box to write your comment about the book"
				onChange={e => setCommentText(e.target.value)}
				className="bg-white text-black placeholder-black textarea textarea-bordered w-[95vw] sm:w-1/2 h-30 focus:outline-accent-focus"
				placeholder="What do you think about this book?"
			></textarea>
			<button
				id="commentSubmitButton"
				onClick={() => {
					commentText.length > 0 ? postComment() : alert("Cannot post empty comment!");
				}}
				className="mt-5 btn btn-accent"
			>
				Submit
			</button>
			{show ? (
				<div
					id="feedbackPopup"
					className="alert alert-success shadow-lg fixed top-3 right-3 w-auto"
				>
					<div>
						<svg
							xmlns="http://www.w3.org/2000/svg"
							className="stroke-current flex-shrink-0 h-6 w-6"
							fill="none"
							viewBox="0 0 24 24"
						>
							<path
								strokeLinecap="round"
								strokeLinejoin="round"
								strokeWidth="2"
								d="M9 12l2 2 4-4m6 2a9 9 0 11-18 0 9 9 0 0118 0z"
							/>
						</svg>
						Your comment has been submitted!
					</div>
				</div>
			) : null}
		</>
	);
}
