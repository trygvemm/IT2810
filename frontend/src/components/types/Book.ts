export default interface Book {
  id: string;
  author: string;
  title: string;
  year: number;
  language: string;
  comments: [Comment];
}
