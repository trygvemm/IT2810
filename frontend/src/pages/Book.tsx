import { gql, useQuery } from "@apollo/client";
import BookInfoDisplay from "../components/BookInfoDisplay";
import CommentInputSection from "../components/CommentInputSection";
import CommentSection from "../components/CommentSection";
import { bookInfo, bookSiteInfo } from "../components/types/bookInfo";

export const GET_BOOK = gql`
	query Query($id: String!) {
		bookId(id: $id) {
			id
			author
			title
			year
			language
			imageLink
			link
			pages
			country
			comments {
				text
				id
			}
		}
	}
`;

const Book = (props: { id: string | null }) => {
	let id = props.id;
	if (id != null) {
		window.sessionStorage.setItem("sessionId", id);
	} else {
		id = window.sessionStorage.getItem("sessionId");
	}
	const { loading, error, data } = useQuery<bookSiteInfo>(GET_BOOK, {
		variables: { id: id },
	});

	if (loading) return <p>Loading...</p>;

	if (error) {
		return <p>Error:(</p>;
	}

	const routeChange = () => {
		window.location.href = "/";
	};

	return (
		<>
			{data?.bookId.map((book: bookInfo) => (
				<div key={book.id}>
					<header role={"banner"}>
						<button
							id="homeButton"
							onClick={() => routeChange()}
							className="btn btn-outline btn-accent sm:absolute left-5 top-5"
						>
							Home
						</button>
						<h1 id="title" aria-label="Book-Title" className="text-center text-6xl p-5 text-accent">
							{book.title}
						</h1>
						<p className="text-lg text-center">Author: {book.author}</p>
					</header>
					<div role={"main"} className="flex flex-col mt-32 items-center">
						<BookInfoDisplay
							year={book.year}
							pages={book.pages}
							country={book.country}
							language={book.language}
						></BookInfoDisplay>
						<a
							id="wikipedia"
							className="text-lg hover:underline text-primary mt-11 focus:outline-accent-focus"
							href={book.link}
						>
							Read more at Wikipedia
						</a>
						<div className="flex flex-col items-center w-full mt-20 mb-20">
							<CommentInputSection id={id} refetch={GET_BOOK} />

							<h2 className="mt-14 text-2xl mb-5">Comment Section</h2>
							<CommentSection comments={book.comments}></CommentSection>
						</div>
					</div>
				</div>
			))}
		</>
	);
};

export default Book;
