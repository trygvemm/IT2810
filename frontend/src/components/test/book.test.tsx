/* eslint-disable no-restricted-globals */
/* eslint-disable testing-library/render-result-naming-convention */
/* eslint-disable testing-library/no-render-in-setup */
/* eslint-disable @typescript-eslint/no-unused-expressions */
import Book, { GET_BOOK } from "../../pages/Book";
import { screen, render } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import userEvent from "@testing-library/user-event";
import { act } from "react-dom/test-utils";
import { CREATE_COMMENT } from "../CommentInputSection";

const mocks = [
	{
		request: {
			query: GET_BOOK,
			variables: {
				id: null,
			},
		},
		result: {
			data: {
				bookId: [
					{
						id: "1",
						author: "J. K. Rowling",
						country: "England",
						imageLink:
							"https://en.wikipedia.org/wiki/Harry_Potter_and_the_Philosopher%27s_Stone#/media/File:Harry_Potter_and_the_Philosopher's_Stone_Book_Cover.jpg",
						language: "english",
						link: "https://en.wikipedia.org/wiki/Harry_Potter_and_the_Philosopher%27s_Stone",
						pages: 250,
						title: "Harry Potter and the Philosopher's Stone",
						year: 1997,
						comments: [
							{
								text: "Veldig fin bok!",
								id: "1",
							},
						],
					},
				],
			},
		},
	},
	{
		request: {
			query: CREATE_COMMENT,
			variables: {
				text: "En sykt bra bok",
				bookId: null,
			},
		},
		result: {
			data: {
				createComment: [
					{
						text: "En sykt bra bok",
						bookId: null,
					},
				],
				text: "En sykt bra bok",
				bookId: null,
			},
		},
	},
];

describe("Book", () => {
	beforeEach(() => {
		render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<Book id={null} />
			</MockedProvider>
		);
	}),
		test("render", async () => {
			expect(await screen.findByText("Loading...")).toBeInTheDocument();
			expect(screen.queryByText("Error:(")).not.toBeInTheDocument();
			expect(await screen.findByText("Author: J. K. Rowling")).toBeInTheDocument();
			expect(await screen.findByText("Veldig fin bok!")).toBeInTheDocument();
		});

	test("homeclick", async () => {
		await act(async () => {
			await new Promise(resolve => setTimeout(resolve, 0));
		});
		userEvent.click(screen.getByText("Home"));
		expect(location.pathname).toBe("/");
	});

	test("Snapshot", async () => {
		const tree = render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<Book id={null} />
			</MockedProvider>
		);
		await act(async () => {
			await new Promise(resolve => setTimeout(resolve, 0));
		});
		expect(tree).toMatchSnapshot();
	});

	test("Submit", async () => {
		await act(async () => {
			await new Promise(resolve => setTimeout(resolve, 0));
		});
		userEvent.type(
			screen.getByPlaceholderText("What do you think about this book?"),
			"En sykt bra bok"
		);
		expect(screen.getByPlaceholderText("What do you think about this book?")).toHaveValue(
			"En sykt bra bok"
		);
		userEvent.click(screen.getByText("Submit"));
		await act(async () => {
			await new Promise(resolve => setTimeout(resolve, 0));
		});
	});
});
