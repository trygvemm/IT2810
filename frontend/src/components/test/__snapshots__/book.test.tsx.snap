// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <header
          role="banner"
        >
          <button
            class="btn btn-outline btn-accent sm:absolute left-5 top-5"
            id="homeButton"
          >
            Home
          </button>
          <h1
            aria-label="Book-Title"
            class="text-center text-6xl p-5 text-accent"
            id="title"
          >
            Harry Potter and the Philosopher's Stone
          </h1>
          <p
            class="text-lg text-center"
          >
            Author: 
            J. K. Rowling
          </p>
        </header>
        <div
          class="flex flex-col mt-32 items-center"
          role="main"
        >
          <section
            class="text-white stats shadow min-w-[60vw] flex flex-col sm:flex-row"
          >
            <div
              class="stat place-items-center"
            >
              <figure
                class="stat-figure text-secondary w-[32px] h-[32px]"
              >
                <img
                  alt="icon of a calendar"
                  class="object-scale-down"
                  src="/year.svg"
                />
              </figure>
              <h2
                class="stat-title"
              >
                Year
              </h2>
              <p
                class="stat-value"
              >
                1997
              </p>
              <small
                class="stat-desc"
              >
                A modern piece of literature
              </small>
            </div>
            <div
              class="stat place-items-center"
            >
              <figure
                class="stat-figure text-secondary  w-[32px] h-[32px]"
              >
                <img
                  alt="icon of pages"
                  class="object-scale-down"
                  src="/pages.svg"
                />
              </figure>
              <h2
                class="stat-title"
              >
                Pages
              </h2>
              <p
                class="stat-value"
              >
                250
              </p>
              <small
                class="stat-desc"
              >
                A light read!
              </small>
            </div>
            <div
              class="stat place-items-center"
            >
              <figure
                class="stat-figure text-secondary  w-[32px] h-[32px]"
              >
                <img
                  alt="icon of speech bubbles"
                  class="object-scale-down"
                  src="/language.svg"
                />
              </figure>
              <h2
                class="stat-title"
              >
                Language
              </h2>
              <p
                class="stat-value"
              >
                english
              </p>
            </div>
            <div
              class="stat place-items-center"
            >
              <figure
                class="stat-figure text-secondary  w-[32px] h-[32px]"
              >
                <img
                  alt="icon of a globe"
                  class="object-scale-down"
                  src="/country.svg"
                />
              </figure>
              <h2
                class="stat-title"
              >
                Country
              </h2>
              <p
                class="stat-value"
              >
                England
              </p>
            </div>
          </section>
          <a
            class="text-lg hover:underline text-primary mt-11 focus:outline-accent-focus"
            href="https://en.wikipedia.org/wiki/Harry_Potter_and_the_Philosopher%27s_Stone"
            id="wikipedia"
          >
            Read more at Wikipedia
          </a>
          <div
            class="flex flex-col items-center w-full mt-20 mb-20"
          >
            <textarea
              aria-label="Text box to write your comment about the book"
              class="bg-white text-black placeholder-black textarea textarea-bordered w-[95vw] sm:w-1/2 h-30 focus:outline-accent-focus"
              id="commentTextArea"
              placeholder="What do you think about this book?"
            />
            <button
              class="mt-5 btn btn-accent"
              id="commentSubmitButton"
            >
              Submit
            </button>
            <h2
              class="mt-14 text-2xl mb-5"
            >
              Comment Section
            </h2>
            <ul
              aria-label="commentsection"
              class="flex flex-col items-center w-full"
            >
              <li
                aria-label="comment"
                class="p-4 m-1 bg-[#B3C5EF] rounded-xl text-black w-[95vw] sm:w-1/2"
              >
                Veldig fin bok!
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div>
        <header
          role="banner"
        >
          <button
            class="btn btn-outline btn-accent sm:absolute left-5 top-5"
            id="homeButton"
          >
            Home
          </button>
          <h1
            aria-label="Book-Title"
            class="text-center text-6xl p-5 text-accent"
            id="title"
          >
            Harry Potter and the Philosopher's Stone
          </h1>
          <p
            class="text-lg text-center"
          >
            Author: 
            J. K. Rowling
          </p>
        </header>
        <div
          class="flex flex-col mt-32 items-center"
          role="main"
        >
          <section
            class="text-white stats shadow min-w-[60vw] flex flex-col sm:flex-row"
          >
            <div
              class="stat place-items-center"
            >
              <figure
                class="stat-figure text-secondary w-[32px] h-[32px]"
              >
                <img
                  alt="icon of a calendar"
                  class="object-scale-down"
                  src="/year.svg"
                />
              </figure>
              <h2
                class="stat-title"
              >
                Year
              </h2>
              <p
                class="stat-value"
              >
                1997
              </p>
              <small
                class="stat-desc"
              >
                A modern piece of literature
              </small>
            </div>
            <div
              class="stat place-items-center"
            >
              <figure
                class="stat-figure text-secondary  w-[32px] h-[32px]"
              >
                <img
                  alt="icon of pages"
                  class="object-scale-down"
                  src="/pages.svg"
                />
              </figure>
              <h2
                class="stat-title"
              >
                Pages
              </h2>
              <p
                class="stat-value"
              >
                250
              </p>
              <small
                class="stat-desc"
              >
                A light read!
              </small>
            </div>
            <div
              class="stat place-items-center"
            >
              <figure
                class="stat-figure text-secondary  w-[32px] h-[32px]"
              >
                <img
                  alt="icon of speech bubbles"
                  class="object-scale-down"
                  src="/language.svg"
                />
              </figure>
              <h2
                class="stat-title"
              >
                Language
              </h2>
              <p
                class="stat-value"
              >
                english
              </p>
            </div>
            <div
              class="stat place-items-center"
            >
              <figure
                class="stat-figure text-secondary  w-[32px] h-[32px]"
              >
                <img
                  alt="icon of a globe"
                  class="object-scale-down"
                  src="/country.svg"
                />
              </figure>
              <h2
                class="stat-title"
              >
                Country
              </h2>
              <p
                class="stat-value"
              >
                England
              </p>
            </div>
          </section>
          <a
            class="text-lg hover:underline text-primary mt-11 focus:outline-accent-focus"
            href="https://en.wikipedia.org/wiki/Harry_Potter_and_the_Philosopher%27s_Stone"
            id="wikipedia"
          >
            Read more at Wikipedia
          </a>
          <div
            class="flex flex-col items-center w-full mt-20 mb-20"
          >
            <textarea
              aria-label="Text box to write your comment about the book"
              class="bg-white text-black placeholder-black textarea textarea-bordered w-[95vw] sm:w-1/2 h-30 focus:outline-accent-focus"
              id="commentTextArea"
              placeholder="What do you think about this book?"
            />
            <button
              class="mt-5 btn btn-accent"
              id="commentSubmitButton"
            >
              Submit
            </button>
            <h2
              class="mt-14 text-2xl mb-5"
            >
              Comment Section
            </h2>
            <ul
              aria-label="commentsection"
              class="flex flex-col items-center w-full"
            >
              <li
                aria-label="comment"
                class="p-4 m-1 bg-[#B3C5EF] rounded-xl text-black w-[95vw] sm:w-1/2"
              >
                Veldig fin bok!
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <header
        role="banner"
      >
        <button
          class="btn btn-outline btn-accent sm:absolute left-5 top-5"
          id="homeButton"
        >
          Home
        </button>
        <h1
          aria-label="Book-Title"
          class="text-center text-6xl p-5 text-accent"
          id="title"
        >
          Harry Potter and the Philosopher's Stone
        </h1>
        <p
          class="text-lg text-center"
        >
          Author: 
          J. K. Rowling
        </p>
      </header>
      <div
        class="flex flex-col mt-32 items-center"
        role="main"
      >
        <section
          class="text-white stats shadow min-w-[60vw] flex flex-col sm:flex-row"
        >
          <div
            class="stat place-items-center"
          >
            <figure
              class="stat-figure text-secondary w-[32px] h-[32px]"
            >
              <img
                alt="icon of a calendar"
                class="object-scale-down"
                src="/year.svg"
              />
            </figure>
            <h2
              class="stat-title"
            >
              Year
            </h2>
            <p
              class="stat-value"
            >
              1997
            </p>
            <small
              class="stat-desc"
            >
              A modern piece of literature
            </small>
          </div>
          <div
            class="stat place-items-center"
          >
            <figure
              class="stat-figure text-secondary  w-[32px] h-[32px]"
            >
              <img
                alt="icon of pages"
                class="object-scale-down"
                src="/pages.svg"
              />
            </figure>
            <h2
              class="stat-title"
            >
              Pages
            </h2>
            <p
              class="stat-value"
            >
              250
            </p>
            <small
              class="stat-desc"
            >
              A light read!
            </small>
          </div>
          <div
            class="stat place-items-center"
          >
            <figure
              class="stat-figure text-secondary  w-[32px] h-[32px]"
            >
              <img
                alt="icon of speech bubbles"
                class="object-scale-down"
                src="/language.svg"
              />
            </figure>
            <h2
              class="stat-title"
            >
              Language
            </h2>
            <p
              class="stat-value"
            >
              english
            </p>
          </div>
          <div
            class="stat place-items-center"
          >
            <figure
              class="stat-figure text-secondary  w-[32px] h-[32px]"
            >
              <img
                alt="icon of a globe"
                class="object-scale-down"
                src="/country.svg"
              />
            </figure>
            <h2
              class="stat-title"
            >
              Country
            </h2>
            <p
              class="stat-value"
            >
              England
            </p>
          </div>
        </section>
        <a
          class="text-lg hover:underline text-primary mt-11 focus:outline-accent-focus"
          href="https://en.wikipedia.org/wiki/Harry_Potter_and_the_Philosopher%27s_Stone"
          id="wikipedia"
        >
          Read more at Wikipedia
        </a>
        <div
          class="flex flex-col items-center w-full mt-20 mb-20"
        >
          <textarea
            aria-label="Text box to write your comment about the book"
            class="bg-white text-black placeholder-black textarea textarea-bordered w-[95vw] sm:w-1/2 h-30 focus:outline-accent-focus"
            id="commentTextArea"
            placeholder="What do you think about this book?"
          />
          <button
            class="mt-5 btn btn-accent"
            id="commentSubmitButton"
          >
            Submit
          </button>
          <h2
            class="mt-14 text-2xl mb-5"
          >
            Comment Section
          </h2>
          <ul
            aria-label="commentsection"
            class="flex flex-col items-center w-full"
          >
            <li
              aria-label="comment"
              class="p-4 m-1 bg-[#B3C5EF] rounded-xl text-black w-[95vw] sm:w-1/2"
            >
              Veldig fin bok!
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
